crates/primitives/snos-output/src/codec.rs

```diff
@@ -0,0 +1,144 @@
+use alloc::vec::Vec;
+
+use starknet_api::hash::StarkFelt;
+
+use crate::felt_reader::{FeltReader, FeltReaderError};
+use crate::{MessageL1ToL2, MessageL2ToL1, StarknetOsOutput};
+
+/// This codec allows to convert structured OS program output into array of felts
+///
+/// In order to prepare parameters for the Starknet contract `updateState` method:
+///     1. Cast the output to dynamic uint256[] array
+///     2. Get onchain data hash & size
+///     3. ABI encode parameters
+///  
+/// NOTE: Field element (252 bit) is encoded as an EVM word (256 bit) and vice versa
+/// EVM developer should be aware of that and prevent data loss by not using the higest 4 bits
+pub trait SnosCodec: Sized {
+    /// Returns number of field elements required to encode current snos output field
+    fn size_in_felts(&self) -> usize;
+    /// Encodes current snos output field as felt array and appends to the result
+    fn encode_to(self, output: &mut Vec<StarkFelt>);
+    /// Tries to decode snos output field given a felt reader instance
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError>;
+    /// Converts structured snos program output into array of field elements
+    fn into_encoded_vec(self) -> Vec<StarkFelt> {
+        let mut output: Vec<StarkFelt> = Vec::with_capacity(self.size_in_felts());
+        self.encode_to(&mut output);
+        output
+    }
+}
+
+impl SnosCodec for StarkFelt {
+    fn size_in_felts(&self) -> usize {
+        1
+    }
+
+    fn encode_to(self, output: &mut Vec<StarkFelt>) {
+        output.push(self);
+    }
+
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError> {
+        input.read()
+    }
+}
+
+impl<T: SnosCodec> SnosCodec for Vec<T> {
+    fn size_in_felts(&self) -> usize {
+        1 + self.iter().map(|elt| elt.size_in_felts()).sum::<usize>()
+    }
+
+    fn encode_to(self, output: &mut Vec<StarkFelt>) {
+        let segment_size = self.size_in_felts() as u64;
+        output.push(segment_size.into());
+        for elt in self.into_iter() {
+            elt.encode_to(output);
+        }
+    }
+
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError> {
+        let mut segment_reader = FeltReader::new(input.read_segment()?);
+        let mut elements: Vec<T> = Vec::new();
+        while segment_reader.remaining_len() > 0 {
+            elements.push(T::decode(&mut segment_reader)?);
+        }
+        Ok(elements)
+    }
+}
+
+impl SnosCodec for MessageL2ToL1 {
+    fn size_in_felts(&self) -> usize {
+        3 + self.payload.len()
```

Comment:
But the size of a `StarkFelt` counted in felt, will always be 1.
It is its own unit of measure.
Am I missing something?

Reply:
Okay. I get that. We can keep hardcoded `1` for the len of vecs, because those won't change. But you are right about the fields.
Good catch!

Reply:
you are 100% right!

Reply:
done

Reply:
I kept hardcoded `3` or `5` for fields that are felts. Same for a vec of felt: i use it's length rather than iterating over and calling size_in_felt

Reply:
But doesn't that beat the purpose again? The idea is to use size_in_felts() wherever possible. You don't need to iterate, just call size_in_felts() (which would internally iterate but thats abstracted from you)

Reply:
It should not iterate. This method is to be used for optimization purposes. To avoid relocation as your buffer grows. Best, case scenario you get the exact predicted size, you allocate with this capacity, and no reallocation happens.

The worst, case scenario is that you create the default vec with no capacity, and relocate a few times as you write potentially thousands and thousands of bytes.

Here we will end up in the middle.

But iterating over a vec is potentially wayyyy more expensive than doing a reallocation. So we don't want to iterate under any condition, it would beat the whole purpose of doing this estimation.
We prefer a not perfect but okayish that will help us in most cases, than the worst-case scenario. Not perfect but still useful.

Reply:
The way your are supposed to use it is the following:
```rust
let data_to_encode = SomeType::new();
let buffer: Vec<StarkFelt> = Vec::new(data_to_encode::size_in_felt());
data_to_encode.encode(buffer);
```

If you replace the second line by `Vec::new()`, without having a hint of the encoded size, the os will reallocate multiple times.
For 100 elements, there will be 5 relocations. Relocation is one of the most expensive operations there is for a program.

Here is a snippet to help you visualize: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c973a9fcc1354167a95f315a5b063bae

Now if you allocate use `with_capcity`. You will only allocate once if you have the perfect estimation.
If you don't you will probably have to relocate a few times, but you are already sparing some useless computation..

But all this optimization is wasted if you have to iterate over your elements. It is overall still cheaper to reallocate, than to iterate over an array (for struct that are not too small).